RFC: LWC server runtime proposal (#23)
salesforce/lwc-rfcs: text/0112-server-engine.md

When using a type-safe language like TypeScript, it is possible to ensure that all the rendering APIs are fulfilled in all environments at compile-time. In addition, injecting this indirection at runtime ensures that component code doesn't have access to actual DOM APIs when running on the server.

From a drawback perspective, introducing this layer of indirection between the core logic and underlying APIs might introduce a performance overhead and increase the overall size of the engine.

To inject the rendering APIs depending on the environment, we would need to create a different entry point per environment. Below you can find some pseudo-code for each entry point. The two entry points use the same `createComponent` abstraction to create an LWC component but pass different renderers specific to the target environment.

> **`entry-points/dom.js`:**

Reviewer: ```suggestion
**`entry-points/dom.ts`:**
```

```ts
import { LightningElement, createComponent } from '@lwc/engine-core';

const domRenderer = {

>     createElement(tagName: string): Element {

Reviewer: super tiny: you have this as a file called `dom.js` but it is a typescript file :)

        return document.createElement(tagName);
    },
    insertBefore(
        node: Node,
        parent: Element,
        anchor: Node
    ): void {
        return parent.insertBefore(node, anchor);
    },
};

// The method dedicated to create an LWC component in a dom environment. It returns a DOM element
// that can then inserted into the document to render.
export function createElement(
    tagName: string,
    opts: { Ctor: typeof LightningElement },
): Element {
    const elm = document.createElement(tagName);
    createComponent(elm, Ctor, domRenderer);
    return elm;
}
```

> **`entry-points/server.js`:**

Reviewer: ```suggestion
**`entry-points/server.ts`:**
```

**No asynchronous operations allowed:** This also means that if a component needs to do an asynchronous operation to fetch data, it will not be possible to do so when rendered on the server. All the asynchronous data dependencies for a component subtree needs to be retrieved prior to rendering the component and needs to be passed via public properties for it to be rendered. Without adding new primitives to the LWC framework, updating the state of a component asynchronously would violate the previous constraints. Other popular UI frameworks are currently working on supporting asynchronous rendering and integrating it into their current SSR solution, but it is a complex feature to implement.

**The `renderedCallback` lifecycle hook will not execute on the server:** When running in a browser, this hook is the first life cycle hook which gives the component author access to the rendered DOM elements. If the component were to attempt to access those APIs on the server it would result in a runtime error since the DOM APIs are not mocked.

> **Wire adapters will not be invoked:** The Wire protocol emits a stream of data to a component. The current protocol doesn't define a way today to indicate that the stream is done emitting new data. Because of this, the first version of SSR will not invoke any wire adapter. The protocol will need to be changed and new primitives will need to be added to LWC to make wire adapters compatible with SSR.

Reviewer 2: We should consider adding an option to capture a one time snapshot of the stream. Disabling wire adapters during SSR  eliminates the ability to use wire protocol to inject initial state including localized text in the SSR output.

Author: As of today, the invocation timing of the `update` callback is different if the wire adapter [has dynamic parameters or not](https://github.com/salesforce/lwc/blob/e1d7efa3e3d20881f8f11caba623492cd6e6addb/packages/%40lwc/engine-core/src/framework/wiring.ts#L336-L340). This difference makes it impossible to invoke wire adapters in a reliable fashion.

Changing the current invocation timing is an observable change that might result in breaking existing customers on the platform. We decided to set wire adapters aside from the initial SSR implementation but it is something that we will reconsider in the future.

Reviewer 3: Did you thought about creating a new hook that fetches data on the server, before serving the component? Nuxt did that, it is called `fetch`. I believe it would be something really great for lwc. With a `fetch` method, components could be served straight away with the data developers need from Salesforce. Pages would be rendered so much faster, not just because the component was rendered on the server, but also because components would no longer need to wait for the wiredMethod to be added to the stack call on the client, which can be slow depending on the specs of the computer, and how many other wiredMethods were added to the stack.

Reviewer 2: Perhaps this can enabled as an optional configuration param when we call `renderLwcToString` or something like that. By default wire adapters can be skipped, but users can opt-in if their use case is deterministic like in my case, I don't connect wire adapters to network requests, they are simply connected to redux style stores to decouple state and UI.

Reviewer 2: Perhaps a reasonable compromise would be to support Wire adapters without dynamic parameters during SSR, which would cover several use-cases (including my own).

Author: IMO, we should make wire execution predictable (both on the client and on the server). It would be surprising that for a developer to discover that a wire stops getting executed in SSR after adding a dynamic param. 

Over the last 6 months, the LWC team has been investigating how we could roll out potentially risky changes like this one (changing the invocation timing of wires) without breaking existing customers in production. I hope that we could leverage one of the approaches we are discussing to solve this issue and enable wire adapter execution for SSR. 

/cc @nolanlawson

This method returns the serialized HTML content rendered by the component. The serialization format is out of the scope of this proposal and will be covered in a different RFC. However, in the first version of the `@lwc/engine-server`, the `renderComponent` will produce an HTML string matching the [declarative shadow DOM proposal](https://github.com/mfreed7/declarative-shadow-dom/blob/master/README.md) format.

#### Component authoring format

The existing `lwc` package stays untouched and will be used to distribute the different versions of the engine. From the developer perspective, the experience writing a component remains identical. Since the LWC engine exposes different APIs depending on the environment, the application owner will be in charge of creating a different entry point for each environment.

**`c/app/app.js`:**

```js
import { LightningElement } from 'lwc';

export default class App extends LightningElement {}
```

**`client.js`:**

```js
import { createElement } from 'lwc'; // Resolves to `@lwc/engine-dom`
import App from 'c/app';

const app = createElement('c-app', { is: App });
document.body.appendChild(app);
```

**`server.js`:**

```js

> import { renderComponent } from '@lwc/engine-server'; // Resolves to `@lwc/engine-server`

Reviewer: ```suggestion
import { renderComponent } from 'lwc'; // Resolves to `@lwc/engine-server`
```
You mentioned in your presentation this was a typo, it's meant to be `lwc` automatically resolves to `@lwc/engine-server`.
